File: bench/python/bench_optim.py
#!/usr/bin/env python3
"""Single-objective optimization benchmarks (B3) — Python side.

Compares scipy.optimize, cma, and pyswarms against hanalyze on a fixed
test-function suite. 30 seeds, median time + final f, success rate.
"""

from __future__ import annotations

import csv
import time
import warnings
from dataclasses import dataclass
from pathlib import Path
from statistics import median

import numpy as np
import scipy.optimize as opt

warnings.filterwarnings("ignore")


REPO = Path(__file__).resolve().parents[2]
OUT = REPO / "bench" / "results" / "python"
OUT.mkdir(parents=True, exist_ok=True)


# ---------------------------------------------------------------------------
# Test functions
# ---------------------------------------------------------------------------

def rosenbrock(x):
    x = np.asarray(x)
    return float(np.sum(100 * (x[1:] - x[:-1] ** 2) ** 2 + (1 - x[:-1]) ** 2))


def rastrigin(x):
    x = np.asarray(x)
    return float(10 * len(x) + np.sum(x ** 2 - 10 * np.cos(2 * np.pi * x)))


def sphere(x):
    x = np.asarray(x)
    return float(np.sum(x ** 2))


def ackley(x):
    x = np.asarray(x)
    n = len(x)
    s1 = float(np.sum(x ** 2) / n)
    s2 = float(np.sum(np.cos(2 * np.pi * x)) / n)
    return -20 * np.exp(-0.2 * np.sqrt(s1)) - np.exp(s2) + 20 + np.e


def levy(x):
    x = np.asarray(x)
    w = 1 + (x - 1) / 4
    term1 = float(np.sin(np.pi * w[0]) ** 2)
    term2 = float(np.sum((w[:-1] - 1) ** 2 *
                         (1 + 10 * np.sin(np.pi * w[:-1] + 1) ** 2)))
    term3 = float((w[-1] - 1) ** 2 * (1 + np.sin(2 * np.pi * w[-1]) ** 2))
    return term1 + term2 + term3


def griewank(x):
    x = np.asarray(x)
    s = float(np.sum(x ** 2)) / 4000
    p = float(np.prod(np.cos(x / np.sqrt(np.arange(1, len(x) + 1)))))
    return s - p + 1


def schwefel(x):
    x = np.asarray(x)
    d = len(x)
    return 418.9829 * d - float(np.sum(x * np.sin(np.sqrt(np.abs(x)))))


TESTS = [
    ("Rosenbrock_2D", 2, rosenbrock),
    ("Rosenbrock_10D", 10, rosenbrock),
    ("Rastrigin_10D", 10, rastrigin),
    ("Sphere_30D", 30, sphere),
    ("Ackley_10D", 10, ackley),
    ("Levy_10D", 10, levy),
    ("Griewank_10D", 10, griewank),
    ("Schwefel_5D", 5, schwefel),
]


# ---------------------------------------------------------------------------
# Algorithms
# ---------------------------------------------------------------------------

def algo_nelder(f, d, rng):
    x0 = rng.uniform(-2.0, 2.0, d)
    t0 = time.perf_counter()
    r = opt.minimize(f, x0, method="Nelder-Mead",
                     options={"maxiter": 5000, "xatol": 1e-8, "fatol": 1e-8})
    return float(r.fun), 1000.0 * (time.perf_counter() - t0)


def algo_lbfgs(f, d, rng):
    x0 = rng.uniform(-2.0, 2.0, d)
    t0 = time.perf_counter()
    r = opt.minimize(f, x0, method="L-BFGS-B",
                     options={"maxiter": 1000, "ftol": 1e-10})
    return float(r.fun), 1000.0 * (time.perf_counter() - t0)


def algo_de(f, d, rng):
    bounds = [(-5.0, 5.0)] * d
    t0 = time.perf_counter()
    r = opt.differential_evolution(f, bounds, seed=int(rng.integers(2**31)),
                                   maxiter=200, tol=1e-7, polish=False)
    return float(r.fun), 1000.0 * (time.perf_counter() - t0)


def algo_cmaes(f, d, rng):
    import cma
    x0 = rng.uniform(-2.0, 2.0, d)
    sigma = 0.5
    t0 = time.perf_counter()
    es = cma.CMAEvolutionStrategy(
        x0, sigma,
        {"verbose": -9, "maxiter": 200, "tolfun": 1e-8,
         "seed": int(rng.integers(2**31))},
    )
    es.optimize(f)
    return float(es.result.fbest), 1000.0 * (time.perf_counter() - t0)


def algo_sa(f, d, rng):
    bounds = [(-5.0, 5.0)] * d
    t0 = time.perf_counter()
    r = opt.dual_annealing(f, bounds, seed=int(rng.integers(2**31)),
                           maxiter=200)
    return float(r.fun), 1000.0 * (time.perf_counter() - t0)


def algo_pso(f, d, rng):
    import pyswarms.single.global_best as gb
    bounds = (np.full(d, -5.0), np.full(d, 5.0))
    options = {"c1": 0.5, "c2": 0.3, "w": 0.9}
    optr = gb.GlobalBestPSO(n_particles=30, dimensions=d,
                            options=options, bounds=bounds)
    t0 = time.perf_counter()
    cost, _ = optr.optimize(lambda xs: np.array([f(x) for x in xs]),
                            iters=200, verbose=False)
    return float(cost), 1000.0 * (time.perf_counter() - t0)


ALGOS = [
    ("NelderMead", algo_nelder),
    ("LBFGS",      algo_lbfgs),
    ("DE",         algo_de),
    ("CMAES",      algo_cmaes),
    ("SA",         algo_sa),
    ("PSO",        algo_pso),
]


# ---------------------------------------------------------------------------

N_SEEDS = 30
SUCCESS_THR = 1e-2


def main() -> int:
    rows = []
    for fname, d, f in TESTS:
        for aname, algo in ALGOS:
            fs, ts = [], []
            for s in range(N_SEEDS):
                rng = np.random.default_rng(s + hash(aname + fname) % 1000000)
                try:
                    fv, ms = algo(f, d, rng)
                except Exception as e:
                    fv, ms = float("inf"), 0.0
                fs.append(fv)
                ts.append(ms)
            med_f = median(fs)
            med_t = median(ts)
            succ = sum(1 for v in fs if abs(v) < SUCCESS_THR) / N_SEEDS
            rows.append({
                "name": f"{fname}/{aname}",
                "time_ms": med_t,
                "acc_main": med_f,
                "acc_aux": succ,
                "extra": f"median over {N_SEEDS} seeds",
            })

    out = OUT / "optim.csv"
    with open(out, "w", newline="") as fh:
        w = csv.writer(fh)
        w.writerow(["system", "suite", "name",
                    "time_ms", "acc_main", "acc_aux", "extra"])
        for r in rows:
            w.writerow(["python", "optim", r["name"],
                        f"{r['time_ms']:.6g}", f"{r['acc_main']:.6g}",
                        f"{r['acc_aux']:.6g}", r["extra"]])
    print(f"wrote {len(rows)} rows -> {out}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: bench/haskell/BenchOptim.hs
{-# LANGUAGE OverloadedStrings #-}
{-# OPTIONS_GHC -fno-full-laziness -fno-cse #-}
-- | Single-objective optimization benchmarks (B3).
--
-- Each (algorithm × test function) combination is run for 30 seeds and we
-- record:
--   - median wall time per run (ms),
--   - median final objective value @f(x*)@,
--   - success rate: @|f(x*)| < 1e-2@ (Sphere/Ackley/Levy: known optimum 0;
--     Rosenbrock: known min 0; Rastrigin: known min 0).

module Main where

import qualified System.Random.MWC       as MWC
import           Data.List               (sort, minimumBy)
import           Data.Ord                (comparing)
import           Control.Monad           (forM)

import qualified Optim.NelderMead        as NM
import qualified Optim.LBFGS             as LB
import qualified Optim.LineSearch        as LS
import qualified Optim.DifferentialEvolution as DE
import qualified Optim.CMAES             as CM
import qualified Optim.SimulatedAnnealing    as SA
import qualified Optim.ParticleSwarm     as PS
import qualified Optim.Common            as OC

import           BenchUtil

-- ---------------------------------------------------------------------------
-- Test functions (all minimization, true optimum f(x*) = 0)
-- ---------------------------------------------------------------------------

rosenbrock, rastrigin, sphere, ackley, levy :: [Double] -> Double
rosenbrock xs =
  sum [ 100 * (xs!!(i+1) - xs!!i ** 2)^(2::Int)
      + (1 - xs!!i)^(2::Int)
      | i <- [0 .. length xs - 2] ]
rastrigin xs =
  10 * fromIntegral (length xs)
    + sum [x*x - 10 * cos (2 * pi * x) | x <- xs]
sphere = sum . map (^(2::Int))
ackley xs =
  let n  = fromIntegral (length xs) :: Double
      s1 = sum (map (^(2::Int)) xs) / n
      s2 = sum (map (\x -> cos (2*pi*x)) xs) / n
  in - 20 * exp (- 0.2 * sqrt s1) - exp s2 + 20 + exp 1
levy xs =
  let w i = 1 + (xs!!i - 1) / 4
      d   = length xs
      sumMid = sum [ (w i - 1)^(2::Int) * (1 + 10 * sin (pi * w i + 1)^(2::Int))
                   | i <- [0 .. d - 2] ]
  in sin (pi * w 0)^(2::Int)
   + sumMid
   + (w (d-1) - 1)^(2::Int) * (1 + sin (2 * pi * w (d-1))^(2::Int))

-- | Griewank function: smooth multimodal with global min at 0.
-- f(x) = sum(x_i^2)/4000 - prod(cos(x_i/sqrt(i+1))) + 1
griewank :: [Double] -> Double
griewank xs =
  let s = sum (map (^(2::Int)) xs) / 4000
      p = product [ cos (x / sqrt (fromIntegral i))
                  | (i, x) <- zip [1 :: Int ..] xs ]
  in s - p + 1

-- | Schwefel function: very deceptive, global min near boundary at
-- x_i = 420.9687, f* = 0. Hard for local optimizers.
-- f(x) = 418.9829 d - sum(x_i sin(sqrt|x_i|))
schwefel :: [Double] -> Double
schwefel xs =
  let d = length xs
  in 418.9829 * fromIntegral d
   - sum [ x * sin (sqrt (abs x)) | x <- xs ]

testFns :: [(String, Int, [Double] -> Double)]
testFns =
  [ ("Rosenbrock_2D",  2, rosenbrock)
  , ("Rosenbrock_10D", 10, rosenbrock)
  , ("Rastrigin_10D",  10, rastrigin)
  , ("Sphere_30D",     30, sphere)
  , ("Ackley_10D",     10, ackley)
  , ("Levy_10D",       10, levy)
  , ("Griewank_10D",   10, griewank)
  , ("Schwefel_5D",    5,  schwefel)
  ]

-- ---------------------------------------------------------------------------
-- Adapters: each algorithm returns (final f(x*), wall-time ms)
-- ---------------------------------------------------------------------------

data Algo = Algo
  { algoName :: String
  , algoRun  :: ([Double] -> Double) -> Int -> IO (Double, Double)
                -- ^ given f and dim, returns (f_final, ms)
  }

algoNM, algoLBFGS, algoDE, algoCMA, algoSA, algoPSO :: Algo

algoNM = Algo "NelderMead" $ \f d -> do
  x0 <- initSeed d
  (ms, r) <- timeitIO 1 OC.orValue (\_ -> NM.runNelderMead f x0)
  return (OC.orValue r, ms)

-- | L-BFGS の勾配は中央差分で代用 (Numeric variant)。
algoLBFGS = Algo "LBFGS" $ \f d -> do
  x0 <- initSeed d
  (ms, r) <- timeitIO 1 OC.orValue
               (\_ -> LB.runLBFGSNumeric LB.defaultLBFGSConfig f x0)
  return (OC.orValue r, ms)

algoDE = Algo "DE" $ \f d -> do
  let bs = replicate d (-5.0, 5.0)
  gen <- MWC.createSystemRandom
  (ms, r) <- timeitIO 1 OC.orValue (\_ -> DE.runDE bs f gen)
  return (OC.orValue r, ms)

algoCMA = Algo "CMAES" $ \f d -> do
  x0 <- initSeed d
  gen <- MWC.createSystemRandom
  (ms, r) <- timeitIO 1 OC.orValue (\_ -> CM.runCMAES f x0 gen)
  return (OC.orValue r, ms)

algoSA = Algo "SA" $ \f d -> do
  let bs = replicate d (-5.0, 5.0)
  gen <- MWC.createSystemRandom
  -- P42 (2026-05-07): the previous (20 runs × 10000 iter, LBFGS every
  -- 10 iter) configuration spawned ~20K inner LBFGS refines × ~1000
  -- numeric-gradient f calls each = ~20M f calls and dominated the
  -- ~1900 ms wall-time.
  --
  -- Settled on (10 runs, every 20) which gives ~5K LBFGS refines
  -- (1/4 of the baseline) — 3-7× faster across all benches with
  -- equivalent quality. We did try (5 runs, every 50) but Rastrigin
  -- collapsed to f≈0.99. We also tried (15 runs, every 20): 1.5×
  -- slower for no measurable quality gain. Note: Rastrigin escape
  -- rate fluctuates 23-70% across re-runs because the SA harness
  -- uses MWC.createSystemRandom (non-deterministic seed) — single
  -- 30-seed runs are noisy; the algorithm itself is robust.
  let cfg = (SA.defaultSAConfig bs)
              { SA.saProposal       = SA.Tsallis 2.62
              , SA.saAccept         = SA.Boltzmann
              , SA.saLocalMethod    = SA.LocalLBFGS
              , SA.saLocalEvery     = Just 20
              , SA.saInitTemp       = 5230.0
              , SA.saRestartIfStuck = Nothing
              , SA.saStop           = (SA.saStop (SA.defaultSAConfig bs))
                                        { OC.stMaxIter = 10000 }
              }
      nRuns = 10 :: Int
  (ms, r) <- timeitIO 1 OC.orValue $ \_ -> do
    rs <- mapM (\_ -> do
                   x0 <- initSeed d
                   SA.runSAWith cfg f x0 gen) [1 .. nRuns]
    return (minimumBy (comparing OC.orValue) rs)
  return (OC.orValue r, ms)

algoPSO = Algo "PSO" $ \f d -> do
  let bs = replicate d (-5.0, 5.0)
  gen <- MWC.createSystemRandom
  (ms, r) <- timeitIO 1 OC.orValue (\_ -> PS.runPSO bs f gen)
  return (OC.orValue r, ms)

initSeed :: Int -> IO [Double]
initSeed d = do
  gen <- MWC.createSystemRandom
  OC.sampleUniformIn (replicate d (-2.0, 2.0)) gen

algos :: [Algo]
algos = [algoNM, algoLBFGS, algoDE, algoCMA, algoSA, algoPSO]

-- ---------------------------------------------------------------------------
-- Run loop
-- ---------------------------------------------------------------------------

nSeeds :: Int
nSeeds = 30

successThr :: Double
successThr = 1e-2

main :: IO ()
main = do
  rows <- fmap concat $ forM testFns $ \(fname, d, f) ->
    forM algos $ \alg -> do
      results <- mapM (\_ -> (algoRun alg) f d) [1 .. nSeeds]
      let (fs, ts) = unzip results
          medF  = median fs
          medMs = median ts
          succRate = fromIntegral
                       (length (filter (\v -> abs v < successThr) fs))
                     / fromIntegral nSeeds :: Double
      return $ BenchRow "haskell" "optim"
        (fname ++ "/" ++ algoName alg) medMs medF succRate
        ("median over " ++ show nSeeds ++ " seeds")
  writeRows "bench/results/haskell/optim.csv" rows
  putStrLn $ "wrote " ++ show (length rows)
          ++ " rows → bench/results/haskell/optim.csv"

median :: Ord a => [a] -> a
median xs =
  let s = sort xs
  in s !! (length s `div` 2)
